### ISSUE TITLE
Update APIs to support passing additional conditioning information to the processor's map

### ISSUE THREAD

**Issue by sgreenbury** [alan-turing-institute/autocast#154]
From discussion with @ContiPaolo @radka-j.

### Decription
Currently the `Processor` base class has the following `map`:
https://github.com/alan-turing-institute/autocast/blob/d4725aaa35cf646ecca384bf60dc8e97fd7937cb/src/autocast/processors/base.py#L36
so there is no direct way to support passing additional conditioning information (e.g. constant scalars, fields, boundary conditions etc) here. 

The way this currently is possible with FNOs and ViT is through using e.g. the `PermuteConcat` encoder as part of the `EncoderProcessorDecoder` such that the additional conditioning information is concatenated with `x=input_fields` and then can be passed as a single tensor to the processor.

However, this approach is limiting for the diffusion/flow-matching models where training their backbones can be improved by handling the different tensor types/shapes distinctly. See e.g. [azula UNet backbone AP](https://azula.readthedocs.io/stable/api/azula.nn.unet.html#azula.nn.unet.UNet.forward) that supports distinct conditioning for `mod` embeddings (such as the diffusion time, embedded class labels, etc) and `cond` that have the same shape as the `x_t` to be denoised.

### Possible API update
A discussed update to enhance the support for these models could:
1. Update `Processor`'s `map` to:
```python
def map(self, x: Tensor, mod: Tensor | None = None) -> Tensor: ...
```
or potentially `def map(self, x: Tensor, mod: Tensor | None = None, cond: Tensor | None = None) -> Tensor: ...`

2. Update the base `Encoder` to return an additional return type `tuple[TensorBNC, TensorBNC | None]` that can be used for conditioning:
```python
def encode(self, batch: Batch) -> TensorBNC | tuple[TensorBNC, TensorBNC | None]:
```
since the union is in the return types of the base class, [LSP](https://en.wikipedia.org/wiki/Liskov_substitution_principle) we can recreate the current base class as a subclass of the new base class without breaking the API, with a different subclass for each case, e.g.:
```python
class EncoderBase:
    # ...
    def encode(self, batch: Batch) -> TensorBNC | tuple[TensorBNC, TensorBNC | None]:

class Encoder(EncoderBase):
    # ...
    def encode(self, batch: Batch) -> TensorBNC:

class EncoderWithConditioning(EncoderBase):
    # ...
    def encode(self, batch: Batch) -> tuple[TensorBNC, TensorBNC | None]:
```
 
3. Which encoders can be combined with which `Processor`s would need consideration since ones that return conditioning tensor's would need to have this passed to the processor while ones without would not have anything to pass here and would default to `None` as mentioned in 1. (or directly pass `None` if no default).
### Alternatives
Alternative options include:
- Updating the `Processor`'s `map` API to receive a generic batch like the `loss` currently does, e.g.:
```python
class Processor(ABC, nn.Module, Generic[BatchT]):
    # ...
    def map(self, x: BatchT) -> Tensor:
```
A plus here would be the processor can determine how to combine fields of a given batch type; a minus would be that the output of `.encode()` calls would not be able to be directly passed without constructing into a `BatchT`
- Have `**kwargs`:
```python
class Processor(ABC, nn.Module, Generic[BatchT]):
    # ...
    def map(self, x: Tensor, **kwargs) -> Tensor:
```
A plus would be creating additional flexibility for downstream subclassing beyond a single additional conditioning tensor; a minus is that it is general making it harder to ensure/see compatibility between different parts of the APIs.
